contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -1588,6 +1590,234 @@ create_xp_instance_regread_in_master_dbo_internal(PG_FUNCTION_ARGS)
 	PG_RETURN_INT32(0);
 }
 
+Datum
+sp_execute_postgresql(PG_FUNCTION_ARGS)
+{
+	List	   *parsetree_list;
+	ListCell   *parsetree_item;
+	char	   *extensionStmt;
+	Node	   *stmt;
+	size_t		len;
+	const char *saved_dialect = GetConfigOption("babelfishpg_tsql.sql_dialect", true, true);
+	Oid			current_user_id = GetUserId();
+	const char *saved_buffer = "%s, \"$user\", sys, pg_catalog";
+	const char *new_buffer = "public, %s, \"$user\", sys, pg_catalog";
+	
+	PG_TRY();
+	{
+		SetCurrentRoleId(GetSessionUserId(), false);
+		if (!(superuser() || role_is_sa(GetSessionUserId())))
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+					 errmsg("permission denied to create extension")));
+		}
```

Comment:
Let's not first set the current role to sessionuser. First check whether current statement is extension statement or not. if yes then check whether current sessionuser is superuser and sysadmin.